cmake/MMDeploy.cmake

```diff
@@ -0,0 +1,113 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+
+function (mmdeploy_export NAME)
+    install(TARGETS ${NAME}
+            EXPORT MMDeployTargets
+            ARCHIVE DESTINATION lib
+            LIBRARY DESTINATION lib
+            RUNTIME DESTINATION bin)
+endfunction ()
+
+
+function (mmdeploy_add_library NAME)
+    add_library(${NAME} ${ARGN})
+    target_compile_definitions(${NAME} PRIVATE -DMMDEPLOY_API_EXPORTS=1)
+    get_target_property(_TYPE ${NAME} TYPE)
+    if (_TYPE STREQUAL STATIC_LIBRARY)
+        set_target_properties(${NAME} PROPERTIES POSITION_INDEPENDENT_CODE 1)
+    elseif (_TYPE STREQUAL SHARED_LIBRARY)
+    else ()
+        message(FATAL_ERROR "unsupported type: ${_TYPE}")
+    endif ()
+    target_link_libraries(MMDeployLibs INTERFACE ${NAME})
+    mmdeploy_export (${NAME})
+endfunction ()
+
+
+function (mmdeploy_add_module NAME)
+    cmake_parse_arguments(_MMDEPLOY "EXCLUDE" "" "" ${ARGN})
+    add_library(${NAME} ${_MMDEPLOY_UNPARSED_ARGUMENTS})
+    get_target_property(_TYPE ${NAME} TYPE)
+    target_link_libraries(${NAME} PRIVATE mmdeploy::core)
+    if (_TYPE STREQUAL STATIC_LIBRARY)
+        set_target_properties(${NAME} PROPERTIES POSITION_INDEPENDENT_CODE 1)
+        if (MSVC)
+            target_link_options(${NAME} INTERFACE "/WHOLEARCHIVE:${NAME}")
+        endif ()
+        if (NOT _MMDEPLOY_EXCLUDE)
+            target_link_libraries(MMDeployStaticModules INTERFACE ${NAME})
+        endif ()
+    elseif (_TYPE STREQUAL SHARED_LIBRARY)
+        if (NOT _MMDEPLOY_EXCLUDE)
+            target_link_libraries(MMDeployDynamicModules INTERFACE ${NAME})
+        endif ()
+    else ()
+        message(FATAL_ERROR "unsupported type: ${_TYPE}")
+    endif ()
+    if (NOT _MMDEPLOY_EXCLUDE)
+        mmdeploy_export(${NAME})
+    endif ()
+endfunction ()
+
+
+function (_mmdeploy_flatten_modules RETVAL)
+    set(_RETVAL)
+    foreach (ARG IN LISTS ARGN)
+        get_target_property(TYPE ${ARG} TYPE)
+        message(STATUS "${ARG} ${TYPE}")
+        if (TYPE STREQUAL "INTERFACE_LIBRARY")
+            get_target_property(LIBS ${ARG} INTERFACE_LINK_LIBRARIES)
+            if (LIBS)
+                list(FILTER LIBS EXCLUDE REGEX ::@)
```

Comment:
`LIBS` pattern is different in different version of cmake. 
For example, in cmake v3.14, `LIBS` is like `mmdeploy_cpu_device::@<0x1c3deb0>;`, while in v3.20, it looks like `::@(0x2958920);mmdeploy_cpu_device;::@;`
So, I suggest using an express as `list(TRANSFORM LIBS REPLACE "::@.*;*" ";")`

Reply:
`TRANSFORM` is a per-element operation, the ';' in the pattern does not make much sense. The resulted `LIBS` is full of ';' if the expression is used.